HTML API: Fix void tag nesting with next_token

sirreal opened the pull request:
Trac ticket: Core-60382

When `next_token()` was introduced in #5683, it introduced regression in the HTML Processor whereby void tags remain on the stack of open elements when they shouldn't. This led to invalid values returned from `get_breadcrumbs()`.

The reason was that calling `next_token()` works through a different code path than the HTML Processor runs everything else. To solve this, its sub-classed `next_token()` called `step( self::REPROCESS_CURRENT_TOKEN )` so that the proper HTML accounting takes place.

Unfortunately that same _reprocessing_ code path skipped the step whereby void and self-closing elements are popped from the stack of open elements.

In this patch that step is run through the introduction of a third mode for `step()`, which is `self::PROCESS_CURRENT_TOKEN`. This mode acts as if `self::PROCESS_NEXT_NODE` were called, except it doesn't advance the parser, that already having been done by `next_token()` first.

Follow-up to [57348]

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Use $p variable for processor like other tests

sirreal committed:
Add failing test

sirreal committed:
Update ticket

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessor.php:
```diff
@@ -227,9 +280,9 @@ public function data_void_tags() {
-		$this->assertEquals( $fragment->get_last_error(), WP_HTML_Processor::ERROR_UNSUPPORTED, 'Should have unsupported last error.' );
+		$p = WP_HTML_Processor::create_fragment( '<' . $tag_name . '></' . $tag_name . '>' );
+		$this->assertFalse( $p->next_tag(), 'Should fail to find tag: ' . $tag_name . '.' );
+		$this->assertEquals( $p->get_last_error(), WP_HTML_Processor::ERROR_UNSUPPORTED, 'Should have unsupported last error.' );
```
for new tests I've tried to standardize on `$processor` after receiving feedback that `$p` was unclear for some, confusing it with a `P` element. would be good to do a full rename throughout the tests to unify the examples.

peterwilsoncc replied:
That seems sensible, I reckon it can go against the generic test improvement tickets for 6.5 WP#59647 on trac.

sirreal replied:
Thanks, I'll remove this change and put together a PR with those renames.

sirreal replied:
PR for renames: https://github.com/WordPress/wordpress-develop/pull/5987

dmsnell committed:
Ensure void and self-closing elements pop from stack when advancing.

Previously, the logic to pop void and self-closing elements from the
stack of open elements only ran when stepping into the next node in
a document. With the introduction of `next_token()` there appeared a
new way to reprocesses the current token, so this logic would be skipped
when calling `next_token()` _into_ a void or self-closing element,
leaving it on the stack.

In this patch the logic runs whenever the processor is not reprocessing
the current token. A new class constant communicates that `step()`
should treat the current token as if it arrived there itself, that is,
to process it with the normal rules but without advancing the parser.

sirreal committed:
Revert "Use $p variable for processor like other tests"

This reverts commit 0deab0a60878082c4cf87da10669bdc929e0c891.

sirreal commented:
@dmsnell your fix looks good to me and fixes the test failures I saw in https://github.com/WordPress/wordpress-develop/pull/5794.

👍 This change looks good to me.

dmsnell committed:
Merge branch 'trunk' into html-api/fix-void-nesting-next-token

dmsnell commented:
Merged in [[57507]](https://core.trac.wordpress.org/changeset/57507)
https://github.com/wordpress/wordpress-develop/commit/cdb218b2007693262f2ea2d9717ec0ff55620b80
